Base classification algorithm

Andimeo opened the pull request:
Try to fix #2727 

@anmolshkl Please take a look at the interfaces definition. If nothing goes wrong, I'll start fitting the LDAClassifier into this pattern.

anmolshkl commented on core/domain/classifier/base_classification_algorithm.py:
```diff
@@ -0,0 +1,153 @@
+
+
+class BaseClassificationAlgorithm(object):
+    """A base class for classifiers that uses supervised learning to match
```
@Andimeo I think we should keep the documentation as generic as possible so that in the future we can build other types of classifiers (and not just string classifiers) using the same BaseClassificationAlgorithm.

Andimeo replied:
Sorry I'm not good at phrasing. Can you give me some advice?

anmolshkl replied:
Oh, sure! I think what we can do here is not to put string classifier's documentation here and think of the BaseClassificationAlgorithm class as a template for "any" classifier. For example, in the [load_examples](https://github.com/Andimeo/oppia/blob/14987c24c27cd754d2f2864556e42209259370aa/core/domain/classifier/base_classification_algorithm.py#L140) method, the argument 'examples' can be of the form [data, [labels]] where data can be anything (a string in the case of String classifier or some other data structure alternatively). So you can probably say "_examples: list of 'examples'. Each example is represented by a 2-element list. The first item of the list is the single training data, and the second item is a list of labels that the data should be matched"_. Then in the actual StringClassifier class, you can give a more concrete example like: 
_training_examples = [
    ['i eat fish and vegetables', ['food']],
    ['fish are pets', ['pets']],
    ['my kitten eats fish', ['food', 'pets']]
                ]_
I think the same goes for the other part of documentation as well. We can keep the string classifier specific documentation to the StringClassifer class. The documentation in this class should be a generic indicator of what the derived class should do. Please let me know if you'd like  me to elaborate further.

Andimeo replied:
Oh, I thought we only focus on supervised learning on text-format examples and the descendant classes are only for different models such as SVM, LDA and etc, so the input examples should be in the same format for all concrete implementations.

Do we really need to handle the input examples of other formats? If so, we should have another layer to abstract the common behaviors for a certain input type. For instance, BaseClassificationAlgorithm -> StringClassifier -> LDAStringClassifier.

anmolshkl replied:
In the long term, we are looking to have classifiers for code also. This is something which we don't have to worry about now but our design decisions should take into account the fact that we are not just dealing with text. So at this point what we really have to think about is a structure which can handle multiple algorithms for text classification & multiple algorithms for program (code) classification. So the real hierarchy (just to make things clear) is like: Classifier->(Text Classifier -> (multiple text classifiers), Code Classifier -> (multiple code classifiers)). 

Now, do we really need a separate layer for abstracting the family of string and code classifiers? Well, that's something we need to dig into & see what are the pros/cons of providing the abstraction vs not providing the abstraction. We need to think over this and put our proposal in the design document. I think this is a good design problem.

Also, we need to clearly describe in the design doc about how we will decide what algorithm will be used for classification (the creator shouldn't decide here because this entire process is hidden from him). Secondly, we also need to describe how we will work with models that have been generated using different classification algorithms.

@seanlip would you like to add anything?

PS: @Andimeo If you want a background information on code classification, please take a look [here](https://github.com/oppia/oppia/wiki/Google-Summer-of-Code-2016#applying-machine-learning-to-code-interaction-training-and-generalizing).

Andimeo replied:
@anmolshkl Should I hold on at current stage until we fill the answers to above questions in our design document?

anmolshkl replied:
@Andimeo Yes, please. Let's get the problem and the possible solution into the design doc first.

Andimeo commented:
@anmolshkl Please take a look.

seanlip commented on core/domain/classifier/base_classification_algorithm.py:
```diff
@@ -0,0 +1,92 @@
+
+"""Base class of classification algorithms for free-form text answers"""
+
+from abc import abstractmethod
```
For python standard libraries, do "import abc" and reference using abc.abstractmethod. (That said, see comment below.)

Andimeo replied:
Done

seanlip commented on core/domain/classifier/base_classification_algorithm.py:
```diff
@@ -0,0 +1,92 @@
+          Args:
+              model: A dict representing the classifier.
+        """
+        pass
```
Should these be `raise NotImplementedError` instead? We do that in e.g. https://github.com/oppia/oppia/blob/develop/core/tests/test_utils.py#L145, so I was wondering, what are the pros/cons of abc.abstractmethod over this?

(I ask because we probably want to stick to one pattern for this codebase-wide, so if the latter approach is better we might want to do a global refactor. /cc @BenHenning)

Andimeo replied:
In my opinion, I don't see the difference between `raise NotImplementedError` and `pass`. So probably it's a good idea to stick to the existed pattern.

seanlip replied:
OK, sounds good. Do we still need abc.abstractmethod in that case?

Andimeo replied:
I think `abstractmethod` is necessary. `NotImplementedError` is only raised when the method is invoked. However, if a method is decorated by `abstractmethod`, a `TypeError` will be raised when instantiating the class instance.

You can refer to 'http://stackoverflow.com/questions/4382945/abstract-methods-in-python' for more information.

seanlip commented on core/domain/classifier/lda_string_classfier.py:
```diff
@@ -0,0 +1,598 @@
+# coding: utf-8
```
Shouldn't the equivalent files be deleted in this PR?

Andimeo replied:
Yes, it's not completely done. If the current implementation meets our needs, I'll try to remove the original code and replace corresponding references later.

seanlip replied:
OK, sounds good. Note, btw, that the name of the file is misspelled: classfier --> classifier.

Andimeo replied:
Done

seanlip commented on core/domain/classifier/lda_string_classfier.py:
```diff
@@ -0,0 +1,598 @@
+import copy
+import numpy
+
+from core.domain.classifier.base_classification_algorithm import BaseClassificationAlgorithm #pylint: disable=line-too-long
```
`from core.domain.classifier import base_classification_algorithm`

and refer using base_classification_algorithm.BaseClassificationAlgorithm

That said, these names are a bit long. Maybe instead:

`from core.domain.classifier import base`

and use base.BaseClassifier?

Andimeo replied:
Sure, that's more succinct.

Andimeo commented on core/domain/classifier/base_classification_algorithm.py:
```diff
@@ -17,7 +17,6 @@
 """Base class of classification algorithms for free-form text answers"""
 
```
Done

Andimeo commented on core/domain/classifier/base_classification_algorithm.py:
```diff
@@ -17,7 +17,6 @@
-import copy
 
 
 class BaseClassificationAlgorithm(object):
```
Done

Andimeo commented on core/domain/classifier/base_classification_algorithm.py:
```diff
@@ -26,72 +25,27 @@ class BaseClassificationAlgorithm(object):
     label - An answer group that the doc should correspond to. If a doc is
         being added to train a model, labels are provided. If a doc is being
         added for prediction purposes, no labels are provided. If a doc does
         not match any label, the doc should have only one label, '_default'.
```
Done

Andimeo commented:
Please take another look.

seanlip commented on core/domain/classifier/base.py:
```diff
@@ -0,0 +1,95 @@
+          Args:
+              model: A dict representing the classifier.
+        """
+        pass
```
I thought we were going to raise NotImplementedError for consistency?

Andimeo replied:
Done

seanlip commented:
Hi @anmolshkl -- I can't look in detail now, but I'm happy if you are! Only
one thing: before merging, probably need to remove the existing files with
the same functionality?

On Saturday, December 17, 2016, Anmol <notifications@github.com> wrote:

> *@anmolshkl* commented on this pull request.
>
> @Andimeo <https://github.com/Andimeo> LGTM, @seanlip
> <https://github.com/seanlip> good to merge?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/2743#pullrequestreview-13442603>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKFeyibr0wfVwUnhGhMX4tRsBysuLCV_ks5rI45xgaJpZM4K3gJ6>
> .
>

Andimeo commented:
@anmolshkl @seanlip I've removed `classifier_service` and `classifier_service_test`. In addition, have replaced all the references of the original `StringClassifier` with this `LDAStringClassifier`. PTAL.

anmolshkl commented:
@seanlip what about the failing Travis builds?

seanlip commented:
Perhaps they are flaky? Could try restarting them if the error looks
unrelated to this PR.

On Tuesday, December 20, 2016, Anmol <notifications@github.com> wrote:

> @seanlip <https://github.com/seanlip> what about the failing Travis
> builds?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/2743#issuecomment-268002719>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKFeynTOICYE9nEQFC-B-fp1g-d1rCreks5rJqsfgaJpZM4K3gJ6>
> .
>

anmolshkl commented:
@seanlip the issue is with RUN_LINT and RUN_E2E_TESTS_MISC_PROD. Not sure why lint job would fail when the pre_commit_linter.py has already checked all the files.

seanlip commented:
I restarted the prod test. I think that one's just flaky.

@Andimeo can you look into the lint issue, please? @anmolshkl you can see what's failing by looking at the log [here](https://travis-ci.org/oppia/oppia/jobs/184891673).

anmolshkl commented:
@andimeo Ah the [problem](https://travis-ci.org/oppia/oppia/jobs/184891673#L486)
 is in `oppia/core/tests/performance_tests/string_classifier_performance_test.py`. Since we have modified the StringClassifier class, you'll need to update all its references.

Andimeo commented:
@anmolshkl Thanks for your investigation. This PR is branched from an early commit in which the string_classifier_performance_test had not been added. 

So I rebased upon current develop branch and replaced the corresponding references. PTAL.

anmolshkl commented:
@Andimeo Thanks! Also, do change the year in the licensing header. Let me know, I'll merge it once you are done.

Andimeo commented:
@anmolshkl Done.

anmolshkl commented:
@seanlip there's another e2e test failing, merging is blocked.

seanlip commented:
Seems like it's passed, to me? Note that you can restart flaky tests by clicking on the link to go to Travis, clicking on the test that failed, and clicking the "restart job" button. You'll need to be logged in to Travis (with your github account) otherwise you won't see the "restart job" button.

That should allow it to run again, and if it passes, merging will be enabled.

anmolshkl commented:
@seanlip Thanks! I didn't know about it earlier. I'll merge the branch then.
